### PR TITLE
Auto install on pull

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

- Adds a git hook installer
- Automatically install dependencies whenever developer gets latest 

## Why?

Ensures developer is always up to date with the latest dependencies and Node version

### CODE REVIEW

**Comment by SiAdcock** [comment]
> Can we hold this until make auto loads nvm?

@AWare I think I'd like to get this merged (it's a straight copy-paste from frontend) then consider making improvements later.

As for auto-loading `nvm`, we don't currently prescribe that developers use it, so we can't currently assume that everyone has / wants it. However, I'd prefer to decouple this discussion from this PR; happy to discuss IRL.

**Comment by SiAdcock** [comment]
> I'm not 100% behind installing deps on pull, as about half the times I pull master I have no intention of running that branch.

Interesting. I'd quite like to unpack that, if possible. What would your typical workflow be, after pulling master?

File: scripts/git-hooks/post-merge [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    .shell('git diff-tree -r --name-only --no-commit-id HEAD@{1} HEAD')
+    .then(result => {
+        const changedFiles = result.stdout.split('\n');
+
+        if (changedFiles.includes('.nvmrc')) {
+            const nvmrcVersion = fs.readFileSync('.nvmrc');
+
+            console.log(
+                `${chalk.red(
+                    `Frontend now requires Node v${nvmrcVersion}. Switch to the latest version using \`nvm install\` or download from nodejs.org. Then run \`make reinstall\`.`
```
Comment (sndrs) [inline scripts/git-hooks/post-merge]:
```suggestion
                    `This application now requires Node v${nvmrcVersion}. Switch to the latest version using \`nvm install\` or download from nodejs.org. Then run \`make reinstall\`.`
```

File: scripts/git-hooks/install.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+const rimraf = require('rimraf');
+
+const src = path.resolve(__dirname);
+const target = path.resolve(__dirname, '..', '..', '.git', 'hooks');
+
+// always try and remove any old ones
+try {
+    rimraf.sync(target);
+} catch (e) {
+    /* do nothing */
```
[the remark this answers is not in the document]
Comment (SiAdcock) [reply scripts/git-hooks/install.js]: No, if the `.git/hooks` directory doesn't yet exist, we should let `rimraf.sync` fail silently.

**Comment by SiAdcock** [comment]
Summary of a discussion with @AWare (please let me know if I've missed anything). I believe this change will not break any particular workflow, but it does get new / upgraded dependencies installed as soon as possible.

## If Yarn is not installed or not available in the terminal when getting latest, this will fail

True, but maybe not the worst thing that could happen. If this hook fails, nothing else breaks. And it's better for the developer to find out sooner rather than later that Yarn is required, but not installed.

## I may just be pulling the master branch so I can rebase another branch on top of it

True, but the new branch will still need the latest versions of dependencies to be installed at some point. This just installs them as soon as possible. 

## What if I'm working on a new branch that uninstalls dependencies that are updated on master?

This script would needlessly upgrade the dependency on master, and then after rebasing the dependency would be uninstalled. This is sub-optimal, but a less common scenario and we shouldn't optimise for every eventuality.

Can anyone think of anything important that I've missed? I'm keen to get a firm 👍 or 👎 and draw this to a close.

**Comment by shtukas** [comment]
+1 from me

**Comment by SiAdcock** [comment]
Okay, I'm gonna merge this, let me know if it causes any problems

**Pull request merged**